File: dotnet/Step3CreateJob.cs
/*
 * Step3CreateJob.cs  —  STEP 3: Start processing the PDF
 * =======================================================
 * Sends an uploaded file for processing and gets back a job_id.
 * Run:  dotnet run -- step3
 *
 * EDIT NOTHING HERE. Set these in  ../config.json  under "process":
 *   "process": {
 *     "file_id": "<an uploaded file_id>",
 *     "level": 1,                          // 1 or 2
 *     "requires_manual_review": false      // optional — see below
 *   }
 *
 * Manual review (optional):
 *   Set "requires_manual_review": true if you'd like to review and refine the
 *   automated tagging in the web UI before the tagged PDF becomes downloadable.
 *   When enabled, Step 4 will report the job as "AwaitingManualReview" until you:
 *     1. Go to https://app.accessibilityondemand.ai/login (you'll also get an
 *        email when the file is ready to review).
 *     2. Open the batch, select the file, click Review.
 *     3. On the last page of the review, click the Complete button.
 *   After that, run Step 4 again and it will return the download_url.
 *
 * What it saves to data.json:
 *   "job_process": [ { "file_id": "....", "job_id": "....", "status": "Queued" }, ... ]
 */

using System;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

namespace Aod
{
    public static class Step3CreateJob
    {
        public static async Task RunAsync()
        {
            JsonObject cfg = Helper.LoadConfig();
            string apiKey = Helper.ApiKey();
            JsonObject process = Helper.GetObject(cfg, "process");
            string fileId = Helper.GetString(process, "file_id", "").Trim();
            int level = Helper.GetInt(process, "level", 1);

            // Optional. Only accept a real JSON boolean `true` (strict); anything else
            // (missing, false, "true" as a string, etc.) is treated as false so the
            // default behaviour stays "fully automatic, downloadable right away".
            bool requiresManualReview = Helper.GetBool(process, "requires_manual_review", false);

            if (string.IsNullOrEmpty(fileId))
            {
                Console.WriteLine("[X] No file_id given. Set \"process\": {\"file_id\": ...} in config.json "
                    + "(use an uploaded file_id from Step 2).");
                return;
            }

            var payload = new JsonObject
            {
                ["file_id"] = fileId,
                ["level"] = level,
                ["requires_manual_review"] = requiresManualReview,
            };

            string reviewNote = requiresManualReview ? "  (with manual review)" : "";
            Console.WriteLine($"Starting a job for file_id {fileId} at level {level}{reviewNote} ...");
            var response = await Helper.PostAsync(Helper.BaseUrl + "/jobs/", apiKey, payload.ToJsonString());
            var body = await Helper.ShowResponseAsync(response);

            int code = (int)response.StatusCode;
            if (code == 409)
            {
                Console.WriteLine($"\n[Conflict] This is already processed: change the file_id in config.json ({fileId})");
                Helper.LogFileError(fileId, code, "Conflict - file already processed", body);
            }

            if ((code == 200 || code == 201) && body != null)
            {
                var data = body["data"] as JsonObject ?? new JsonObject();
                string jobId = data["job_id"] != null ? Helper.Str(data["job_id"]) : null;

                if (jobId != null)
                {
                    JsonArray jobProcess = Helper.GetArray("job_process");
                    bool exists = false;
                    foreach (var e in jobProcess)
                        if (Helper.Str(e.AsObject()["job_id"]) == jobId) { exists = true; break; }

                    if (!exists)
                    {
                        var entry = new JsonObject
                        {
                            ["file_id"] = fileId,
                            ["job_id"] = jobId,
                            ["status"] = "Queued",
                        };
                        // Keep a small marker on the entry so data.json makes it clear
                        // this job was started with manual review enabled.
                        if (requiresManualReview) entry["requires_manual_review"] = true;
                        jobProcess.Add(entry);
                        Helper.SaveValue("job_process", jobProcess);
                    }
                    Console.WriteLine("\n[OK] Got job_id: " + jobId);
                    if (requiresManualReview)
                        Console.WriteLine("     (manual review enabled — Step 4 will prompt you to review "
                            + "in the UI once tagging finishes)");
                    Console.WriteLine("Next: run  dotnet run -- step4");
                }
                else
                {
                    Console.WriteLine("\n[!] Could not find 'job_id' in the response. "
                        + "Check the printed response above and update the key name.");
                    Helper.LogFileError(fileId, code, "No job_id in job-create response", body);
                }
            }
            else if (code != 409)
            {
                Console.WriteLine("\n[X] Could not start the job. Check the file_id, level, and status code above.");
                Helper.LogFileError(fileId, code, "Could not start job", body);
            }
        }
    }
}

File: java/Step3CreateJob.java
/*
 * Step3CreateJob.java  —  STEP 3: Start processing the PDF
 * =========================================================
 * Sends an uploaded file for processing (tagging) and gets back a job_id.
 *
 * EDIT NOTHING HERE. Set these in  config.json  under "process":
 *   "process": { "file_id": "<an uploaded file_id>", "level": 1 }   // level 1 or 2
 *
 * How to run (Java 11+):
 *   Mac/Linux:  java -cp ".:lib/gson.jar" Step3CreateJob.java
 *   Windows:    java -cp ".;lib\gson.jar" Step3CreateJob.java
 *
 * What it saves to data.json:
 *   "job_process": [ { "file_id": "....", "job_id": "....", "status": "Queued" }, ... ]
 */

import com.google.gson.*;

public class Step3CreateJob {

    public static void main(String[] args) throws Exception {
        JsonObject cfg = AOD.loadConfig();
        String apiKey = AOD.apiKey();
        JsonObject process = AOD.getObject(cfg, "process");
        String fileId = AOD.getString(process, "file_id", "").trim();
        int level = AOD.getInt(process, "level", 1);

        if (fileId.isEmpty()) {
            System.out.println("[X] No file_id given. Set \"process\": {\"file_id\": ...} in config.json "
                    + "(use an uploaded file_id from Step 2).");
            return;
        }

        JsonObject payload = new JsonObject();
        payload.addProperty("file_id", fileId);
        payload.addProperty("level", level);

        System.out.println("Starting a job for file_id " + fileId + " at level " + level + " ...");
        java.net.http.HttpResponse<String> response = AOD.post(AOD.BASE_URL + "/jobs/", apiKey, payload.toString());
        JsonObject body = AOD.showResponse(response);

        int code = response.statusCode();
        if (code == 409) {
            System.out.println("\n[Conflict] This is already processed: change the file_id in config.json (" + fileId + ")");
        }

        if ((code == 200 || code == 201) && body != null) {
            JsonObject data = body.has("data") && body.get("data").isJsonObject()
                    ? body.getAsJsonObject("data") : new JsonObject();
            String jobId = data.has("job_id") && !data.get("job_id").isJsonNull()
                    ? data.get("job_id").getAsString() : null;

            if (jobId != null) {
                JsonArray jobProcess = AOD.getArray("job_process");
                boolean exists = false;
                for (JsonElement e : jobProcess) {
                    JsonObject j = e.getAsJsonObject();
                    if (j.has("job_id") && j.get("job_id").getAsString().equals(jobId)) { exists = true; break; }
                }
                if (!exists) {
                    JsonObject entry = new JsonObject();
                    entry.addProperty("file_id", fileId);
                    entry.addProperty("job_id", jobId);
                    entry.addProperty("status", "Queued");
                    jobProcess.add(entry);
                    AOD.saveValue("job_process", jobProcess);
                }
                System.out.println("\n[OK] Got job_id: " + jobId);
                System.out.println("Next: run  Step4CheckJob.java");
            } else {
                System.out.println("\n[!] Could not find 'job_id' in the response. "
                        + "Check the printed response above and update the key name.");
            }
        } else {
            System.out.println("\n[X] Could not start the job. Check the file_id, level, and status code above.");
        }
    }
}


/*
 * AOD — shared helper used by every step file.
 * You normally do NOT need to edit this. It holds the Base URL, builds the
 * Authorization header, sends requests, reads your values from config.json,
 * and reads/writes data.json.
 *
 * ALL editable values live in  config.json  — you never edit the .java files.
 */
class AOD {
    static final String BASE_URL = "https://staging.api.accessibilityondemand.space/api/v1";

    // Shared config lives in the REPO ROOT (one level up from this language folder).
    static final java.nio.file.Path CONFIG_FILE = java.nio.file.Paths.get("..", "config.json");
    // data.json stays inside THIS language folder.
    static final java.nio.file.Path DATA_FILE = java.nio.file.Paths.get("data.json");
    static final com.google.gson.Gson GSON =
            new com.google.gson.GsonBuilder().setPrettyPrinting().create();
    static final java.net.http.HttpClient CLIENT = java.net.http.HttpClient.newHttpClient();

    // ---------- config.json (the one file you edit) ----------

    static com.google.gson.JsonObject loadConfig() {
        try {
            if (!java.nio.file.Files.exists(CONFIG_FILE)) {
                System.out.println("[X] config.json was not found at ../config.json (the repo root). "
                        + "Run this file from inside the java folder, with config.json in the folder above it.");
                System.exit(1);
            }
            String txt = java.nio.file.Files.readString(CONFIG_FILE);
            return com.google.gson.JsonParser.parseString(txt).getAsJsonObject();
        } catch (Exception e) {
            System.out.println("[X] Could not read config.json (is the JSON valid?): " + e.getMessage());
            System.exit(1);
            return new com.google.gson.JsonObject(); // unreachable
        }
    }

    /** Read the API key, with a friendly error if it's still the placeholder. */
    static String apiKey() {
        String key = getString(loadConfig(), "api_key", "");
        if (key.isEmpty() || key.equals("aod-xxxxxxxxxxx")) {
            System.out.println("[X] Please set your real \"api_key\" in config.json "
                    + "(it is still the placeholder).");
            System.exit(1);
        }
        return key;
    }

    /** Read a String value from a JsonObject, or return the default if missing/null. */
    static String getString(com.google.gson.JsonObject obj, String key, String def) {
        if (obj != null && obj.has(key) && !obj.get(key).isJsonNull()) {
            return obj.get(key).getAsString();
        }
        return def;
    }

    /** Read an int value from a JsonObject, or return the default if missing/null. */
    static int getInt(com.google.gson.JsonObject obj, String key, int def) {
        if (obj != null && obj.has(key) && !obj.get(key).isJsonNull()) {
            try { return obj.get(key).getAsInt(); } catch (Exception ignored) {}
        }
        return def;
    }

    /** Read a nested object (e.g. "process", "report") from config, or empty object. */
    static com.google.gson.JsonObject getObject(com.google.gson.JsonObject obj, String key) {
        if (obj != null && obj.has(key) && obj.get(key).isJsonObject()) {
            return obj.getAsJsonObject(key);
        }
        return new com.google.gson.JsonObject();
    }

    /** Read a String array from config (e.g. "signed_urls"), ignoring blank/placeholder entries. */
    static java.util.List<String> getStringArray(com.google.gson.JsonObject obj, String key) {
        java.util.List<String> out = new java.util.ArrayList<>();
        if (obj != null && obj.has(key) && obj.get(key).isJsonArray()) {
            for (com.google.gson.JsonElement e : obj.getAsJsonArray(key)) {
                if (e == null || e.isJsonNull()) continue;
                String v = e.getAsString().trim();
                if (v.isEmpty() || v.startsWith("https://your-signed-url")) continue;
                out.add(v);
            }
        }
        return out;
    }

    // ---------- HTTP ----------

    static java.net.http.HttpResponse<String> post(String url, String apiKey, String jsonBody)
            throws Exception {
        java.net.http.HttpRequest req = java.net.http.HttpRequest.newBuilder()
                .uri(java.net.URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .header("Content-Type", "application/json")
                .POST(java.net.http.HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();
        return CLIENT.send(req, java.net.http.HttpResponse.BodyHandlers.ofString());
    }

    static java.net.http.HttpResponse<String> get(String url, String apiKey) throws Exception {
        java.net.http.HttpRequest req = java.net.http.HttpRequest.newBuilder()
                .uri(java.net.URI.create(url))
                .header("Authorization", "Bearer " + apiKey)
                .GET()
                .build();
        return CLIENT.send(req, java.net.http.HttpResponse.BodyHandlers.ofString());
    }

    static com.google.gson.JsonObject showResponse(java.net.http.HttpResponse<String> response) {
        System.out.println("Status code: " + response.statusCode());
        try {
            com.google.gson.JsonObject body =
                    com.google.gson.JsonParser.parseString(response.body()).getAsJsonObject();
            System.out.println("Response:");
            System.out.println(GSON.toJson(body));
            return body;
        } catch (Exception e) {
            System.out.println("Response (text): " + response.body());
            return null;
        }
    }

    // ---------- data.json (shared between steps) ----------

    static com.google.gson.JsonObject loadData() {
        try {
            if (!java.nio.file.Files.exists(DATA_FILE)) return new com.google.gson.JsonObject();
            String txt = java.nio.file.Files.readString(DATA_FILE);
            return com.google.gson.JsonParser.parseString(txt).getAsJsonObject();
        } catch (Exception e) {
            return new com.google.gson.JsonObject();
        }
    }

    static void saveValue(String key, com.google.gson.JsonElement value) {
        try {
            com.google.gson.JsonObject data = loadData();
            data.add(key, value);
            java.nio.file.Files.writeString(DATA_FILE, GSON.toJson(data));
            System.out.println("[saved] '" + key + "' was saved to data.json");
        } catch (Exception e) {
            System.out.println("[!] Could not save to data.json: " + e.getMessage());
        }
    }

    static com.google.gson.JsonArray getArray(String key) {
        com.google.gson.JsonObject data = loadData();
        if (data.has(key) && data.get(key).isJsonArray()) {
            return data.getAsJsonArray(key);
        }
        return new com.google.gson.JsonArray();
    }

    static com.google.gson.JsonArray extractDetailBlocks(com.google.gson.JsonObject body) {
        if (body.has("data") && body.get("data").isJsonObject()) {
            com.google.gson.JsonObject data = body.getAsJsonObject("data");
            if (data.has("detail") && data.get("detail").isJsonArray()) {
                return data.getAsJsonArray("detail");
            }
        }
        if (body.has("error") && body.get("error").isJsonObject()) {
            com.google.gson.JsonObject err = body.getAsJsonObject("error");
            if (err.has("details") && err.get("details").isJsonArray()) {
                return err.getAsJsonArray("details");
            }
        }
        return new com.google.gson.JsonArray();
    }
}
